### PR TITLE
Fall 2024 bugfixes

### PR DESCRIPTION
**Pull request by jones-michael-s**
Hello all,

This PR has a few minor bugfixes and an extension to aamod_firstlevel_model:

1) aas_processBIDS -- fixed a tsvread bug where it will calculate a wrong line count if the terminal newline character is missing in the tsv file.

2) washu_surfacerender -- fixed a bug in explicit color scaling.

3) aamod_secondlevel_randomise -- Apple changed their unix "sleep" syntax so I changed file blocking to use Matlab "pause" instead.

4) aamod_firstlevel_model -- added an explicit mask option we need for our work using aa to analyze HDDOT data. HDDOT doesn't provide a structural image, so you need to pass a mask based on the optode array field of view (or use SPM implicit masking, which can be unreliable). Also, I reinstated the physreg stream to the header which somehow got deleted. Note this file is committed twice in the PR -- the first time I used the wrong version and just adding a new commit seemed less error-prone than attempting clever git-fu to delete the first commit...)

5) Finally, I added a new module: aamod_joyplot. This creates a "joyplot" of functional data which we find provides better QA information than a greyplot (aka carpetplot). Perhaps others might find it useful.  Here's an example:

sample_joyplot

The module also plots all outliers in the dataset, which we find helpful to identify artifacts that are correlated across subjects.

### CODE REVIEW

File: aa_modules/aamod_firstlevel_model.xml [hunk 1]
```diff
@@ -36,7 +36,11 @@
 			otherwise (0-1) - fraction threshold
 			-->
             <firstlevelmasking>1</firstlevelmasking>
-            
+
+            <!-- optional explicit mask (fullpath to .nii file) applied to all subjects -->
+            <!-- SPM will combine (intersect) this with firstlevelmasking option -->
+            <explicitmaskfname></explicitmaskfname>
```
Comment (tiborauer) [inline aa_modules/aamod_firstlevel_model.xml:42]: Would not be more c-aa-nonical to add a mask as a stream?

Comment (tiborauer) [reply aa_modules/aamod_firstlevel_model.xml:42]: You can use study-level stream by using `aas_getfile_bystream_multilevel`

**Comment by jones-michael-s** [comment]
Apparently the PR update didn't include the mfile change (!)

I pushed a new PR...

File: aa_modules/aamod_firstlevel_model.xml (hunk 1 above)
Comment (jones-michael-s) [reply aa_modules/aamod_firstlevel_model.xml]: I've looked into aas_getfiles_bystream_multilevel and I don't understand how it solves the problem. 

The examples I can find in the code appear to be modules that can run in different domains (often aliased) and all of them pass the current domain to aas_getfiles_bystream_multilevel (for example, using the variable aap.tasklist.currenttask.domain).

This is not what I need. 

I need the user to be able to specify an explicit mask at the subject level -- for example:

`aap = aas_addinitialstream(aap, 'explicitmask', 'sub-01', 'mask.nii');
`

or at the study level -- for example:

`aap = aas_addinitialstream(aap,'explicitmask','mask.nii');
`

(I suppose a user could specify the mask at the session level, but that seems unlikely)

The problem is aamod_firstlevel_model *always* runs as subject domain. 

So I don't need the domain of the _module_ (e.g., aap.tasklist.currenttask.domain), I need the domain of the _stream_ (so I can pass it to aas_getfiles_bystream_multilevel).

 Is there a function to do that? Or am I not understanding something? (Answer: probably :-)

Comment (tiborauer) [reply aa_modules/aamod_firstlevel_model.xml]: You can still use the _aas_getfiles_bystream_multilevel_ function with the lowest possible stream level (i.e., subject). If it cannot find the stream at that level, it keeps looking at one level up.

It is a slightly hacky workaround, I am afraid, and I actually implemented an explicit stream level independent from the task level in reproa.

**Review by tiborauer — APPROVED** [review]

**Pull request merged**